### PR TITLE
feat(move): fully implement Dragon Darts

### PR DESCRIPTION
**Pull request by Aether21**
## Credit
This PR would not be possible without the [hard work of innerthunder on the Poketernity project](https://github.com/Despair-Games/poketernity/pull/46)! Thank you very much for letting me translate this code to PokeRogue!

## What are the changes the user will see?
Dragon Darts now has "smart" targeting, thereby fully implementing the move (save for a few edge cases)

## Why am I making these changes?
Dragon Darts deserves to be fully realized. Also I wanted to use it for mono Dragon runs.

## What are the changes from a developer perspective?

- `enums/move-target`: New MoveTarget option `DRAGON_DARTS` to implement Dragon Darts' dynamic targeting.
- `phases/move-effect-phase:`
    - `DRAGON_DARTS` move targeting adjusts the move's target for the given strike in double battles in two ways:
        - When a new strike for the move is queued, the target for the next hit is set to the prior target's ally (if on field)
        - If the strike would be predicted to fail its hit check against the base target, the move shifts targets to the base target's ally before applying effects.
    - `hitCheck` and `protectedCheck` now take a simulated argument to stop Dragon Darts from activating effects during its redirection. The logic to apply protection effects has been slightly edited to better support this change.
- `test/moves/dragon_darts.test:` Test suite on Dragon Darts' smart targeting based on innerthunder's PR (with enhancements to use new testing utils/add details)

## How to test the changes?
pnpm run test dragon_darts

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`pnpm test:silent`)
  - [x] Have I created new automated tests (`pnpm test:create`) or updated existing tests related to the PR's changes?
- [x] Have I provided screenshots/videos of the changes (if applicable)?
  - [x] Have I made sure that any UI change works for both UI themes (default and legacy)?

Are there any localization additions or changes? If so:
- [n/a] Has a locales PR been created on the [locales](https://github.com/pagefaultgames/pokerogue-locales) repo?
  - [n/a] If so, please leave a link to it here: 
- [n/a] Has the translation team been contacted for proofreading/translation?

Does this require any changes to the assets folder? If so:
  - [n/a] Has a PR been created on the [assets](https://github.com/pagefaultgames/pokerogue-assets) repo?
    - [n/a] If so, please leave a link to it here:

### CODE REVIEW

File: src/battle-scene.ts [hunk 1]
```diff
@@ -825,6 +826,15 @@ export class BattleScene extends SceneBase {
...
 
+  /**
+   * Returns the Pokemon currently on the field that has a certain battler index, or `undefined` if no such Pokemon exists.
+   * This function is allowed to return non-active (e.g., fainted) Pokemon.
+   * @param battlerIndex - The {@linkcode BattlerIndex} to search for.
+   */
+  public getPokemonByBattlerIndex(battlerIndex: BattlerIndex): Pokemon | undefined {
+    return this.getField().find(p => p.getBattlerIndex() === battlerIndex);
+  }
+
```
Comment (Bertie690) [inline src/battle-scene.ts:836]: This is equivalent to `this.getField()[battlerIndex]` for really weird and fucked up reasons.

In hindsight, a helper function like this would be a decent abstraction over how exactly `BattlerIndex`es work (or don't), so I might steal this and implement it across the codebase. :)

**Comment by Aether21** [comment]
Nothing game-breaking, but a slight visual bug. When both opponents protect, the shield only appears on the ally, but both protect messages play.

https://github.com/user-attachments/assets/1f5eef6e-e081-457b-a4de-1a9224a232bc

File: src/data/moves/move.ts
Comment (Bertie690) [inline src/data/moves/move.ts:1]: I'd prefer to write out the edge cases manually for documentation purposes

File: src/data/battler-tags.ts [hunk 2]
```diff
@@ -1767,8 +1782,10 @@ export class ProtectedTag extends BattlerTag {
     );
   }
 
-  lapse(pokemon: Pokemon, lapseType: BattlerTagLapseType): boolean {
-    if (lapseType === BattlerTagLapseType.CUSTOM) {
+  // lapse(pokemon: Pokemon, lapseType: BattlerTagLapseType): boolean {
+  //   if (lapseType === BattlerTagLapseType.CUSTOM) {
```
Comment (fabske0) [inline src/data/battler-tags.ts]:
```suggestion
```

File: src/data/battler-tags.ts [hunk 3]
```diff
@@ -1784,7 +1801,8 @@ export class ProtectedTag extends BattlerTag {
       return true;
     }
 
-    return super.lapse(pokemon, lapseType);
+    //return super.lapse(pokemon, lapseType);
```
Comment (fabske0) [inline src/data/battler-tags.ts]:
```suggestion
```

File: src/data/battler-tags.ts [hunk 4]
```diff
@@ -1806,19 +1824,23 @@ export abstract class ContactProtectedTag extends ProtectedTag {
...
-    const ret = super.lapse(pokemon, lapseType);
+  apply(pokemon: Pokemon, simulated: boolean, attacker: Pokemon, move: Move): boolean {
+    if (!super.apply(pokemon, simulated, attacker, move)) {
+      return false;
+    }
 
     const moveData = getMoveEffectPhaseData(pokemon);
     if (
-      lapseType === BattlerTagLapseType.CUSTOM
+      //lapseType === BattlerTagLapseType.CUSTOM
```
Comment (fabske0) [inline src/data/battler-tags.ts]:
```suggestion
```

File: src/data/battler-tags.ts [hunk 5]
```diff
@@ -1806,19 +1824,23 @@ export abstract class ContactProtectedTag extends ProtectedTag {
...
+      !simulated
       && moveData
       && moveData.move.doesFlagEffectApply({ flag: MoveFlags.MAKES_CONTACT, user: moveData.attacker, target: pokemon })
     ) {
       this.onContact(moveData.attacker, pokemon);
     }
 
-    return ret;
+    //return ret;
```
Comment (fabske0) [inline src/data/battler-tags.ts]:
```suggestion
```

**Review by fabske0 — COMMENTED** [review]
I guess these can be removed?

File: src/data/moves/move.ts [hunk 6]
```diff
@@ -5615,6 +5615,12 @@ export class MatchUserTypeAttr extends VariableMoveTypeAttr {
   }
 }
 
+/**
+ * Implements smart targeting for the move Dragon Darts
+ */
+
+
+
```
Comment (fabske0) [inline src/data/moves/move.ts]: Also this
```suggestion
```

File: src/phases/move-effect-phase.ts [hunk 7]
```diff
@@ -611,7 +677,7 @@ export class MoveEffectPhase extends PokemonPhase {
     // TODO: Fix lock on / mind reader check.
     if (
       user.getTag(BattlerTagType.IGNORE_ACCURACY)
-      && (user.getLastXMoves().find(() => true)?.targets || []).indexOf(target.getBattlerIndex()) !== -1
+      && (user.getLastXMoves()[0]?.targets ?? []).indexOf(target.getBattlerIndex()) !== -1
     ) {
```
Comment (Bertie690) [inline src/phases/move-effect-phase.ts:643]: This (while better) is still completely borked with instruct.

Please revert as I plan on tackling this in another PR.

File: src/data/moves/move.ts [hunk 8]
```diff
@@ -11689,7 +11700,10 @@ export function initMoves() {
     new AttackMove(MoveId.DRAGON_DARTS, PokemonType.DRAGON, MoveCategory.PHYSICAL, 50, 100, 10, -1, 0, 8)
       .attr(MultiHitAttr, MultiHitType._2)
       .makesContact(false)
-      .partial(), // smart targetting is unimplemented
+      .target(MoveTarget.DRAGON_DARTS)
+      .edgeCase(), 
+      // see `dragon_darts.test.ts` for documented edge cases
+      // when both enemies use protect, both hits go into the ally of the current target
```
Comment (DayKev) [inline src/data/moves/move.ts]:
```suggestion
      // see `dragon_darts.test.ts` for documented edge cases
      // when both enemies use protect, both hits go into the ally of the current target
      .edgeCase(),
```

File: src/phases/move-effect-phase.ts [hunk 9]
```diff
@@ -284,9 +290,34 @@ export class MoveEffectPhase extends PokemonPhase {
       user.turnData.hitsLeft = hitCount.value;
     }
 
+    const hasSmartTargeting = this.canApplySmartTargeting();
+
+    /**
```
Comment (DayKev) [inline src/phases/move-effect-phase.ts]:
```suggestion
    /*
```

File: src/phases/move-effect-phase.ts [hunk 10]
```diff
@@ -284,9 +290,34 @@ export class MoveEffectPhase extends PokemonPhase {
...
+     */
+    if (hasSmartTargeting && user.turnData.hitsLeft % 2 === 1) {
+      const targetAlly = this.getFirstTarget()!.getAlly()!;
+      this.adjustedTargets = [targetAlly.getBattlerIndex()];
+    }
+
+    // Update hit checks for each target
+    this.hitChecks = this.getTargets().map(t => this.hitCheck(t, hasSmartTargeting));
+
+    /**
```
Comment (DayKev) [inline src/phases/move-effect-phase.ts]:
```suggestion
    /*
```

File: src/phases/move-effect-phase.ts [hunk 11]
```diff
@@ -355,6 +386,21 @@ export class MoveEffectPhase extends PokemonPhase {
       return;
     }
 
+    /**
```
Comment (DayKev) [inline src/phases/move-effect-phase.ts]:
```suggestion
    /*
```

File: src/phases/move-effect-phase.ts (hunk 7 above)
Comment (Aether21) [reply src/phases/move-effect-phase.ts:643]: I'd love to revert this but my ability to commit is suddenly broken due to some biome stuff.

File: src/data/battler-tags.ts [hunk 12]
```diff
@@ -1806,19 +1822,21 @@ export abstract class ContactProtectedTag extends ProtectedTag {
    * @param lapseType - The type of lapse to apply. If this is not {@linkcode BattlerTagLapseType.CUSTOM CUSTOM}, no effect will be applied.
    * @returns Whether the tag continues to exist after the lapse.
    */
-  lapse(pokemon: Pokemon, lapseType: BattlerTagLapseType): boolean {
-    const ret = super.lapse(pokemon, lapseType);
+  apply(pokemon: Pokemon, simulated: boolean, attacker: Pokemon, move: Move): boolean {
```
Comment (fabske0) [inline src/data/battler-tags.ts:1826]: This needs it's doc updated

Comment (Aether21) [reply src/data/battler-tags.ts:1826]: Done

**Comment by DarthMDev** [comment]
I'd also be interested in this move being implemented , is there anything I can do to help?

**Comment by SirzBenjie** [comment]
There are minor parity differences with mainline due to how this is implemented (for example, if an ally protects while its partner dies on the first hit, the ally's protect should NOT be triggered).

I think the targets' array should really be set in the move phase, and then it uses the first target that passes the hit check in the MEP, and then it queues the second MEP with the other target in the array (or the same one if only one).

This PR could be improved if it integrated a few of the things I was working on myself. Namely, in the move phase, add a new check for targets below the one for counter:

```ts
  /**
   * If the move's target is not an ally of the user, then add the battler index
   */
  protected resolveSmartTargeting(): void {
    const { targets } = this;
    if (!globalScene.currentBattle.double || targets.length !== 1) {
      return;
    }

    const userIndex = this.pokemon.getBattlerIndex();
    const target = targets[0];
    if (target === BattlerIndex.ENEMY && userIndex !== BattlerIndex.ENEMY_2) {
      targets.push(BattlerIndex.ENEMY_2);
    } else if (target === BattlerIndex.ENEMY_2 && userIndex !== BattlerIndex.ENEMY) {
      targets.push(BattlerIndex.ENEMY);
    } else if (target === BattlerIndex.PLAYER && userIndex !== BattlerIndex.PLAYER_2) {
      targets.push(BattlerIndex.PLAYER_2);
    } else if (target === BattlerIndex.PLAYER_2 && userIndex !== BattlerIndex.PLAYER) {
      targets.push(BattlerIndex.PLAYER);
    }
  }
```

Then, in the move effect phase, we need to make sure the _order_ of targets is preserved, changing `getTargets` to this:

```ts
  public getTargets(): Pokemon[] {
    // Resolve targets to active pokemon while preserving order of targets in
    // the target array. Target order preservation needed for Dragon Darts' smart targeting.
    const candidates = globalScene.getField(true);
    const candidateIndices = candidates.map(p => p.getBattlerIndex());
    const ret: Pokemon[] = [];
    for (const targetIndex of this.targets) {
      const candidateIndex = candidateIndices.indexOf(targetIndex);
      if (candidateIndex !== -1) {
        ret.push(candidates[candidateIndex]);
      }
    }
    return ret;
  }
```

I think this will end up having better parity, if these changes can be integrated with this PR.
